File: DSA/Manav-Rajput/Day56_question2.java
class Solution {
    // Function to sort the given linked list using Merge Sort.
  static Node mergeSort(Node head) {
        // Base case: if head is null or there is only one element
        if (head == null || head.next == null) {
            return head;
        }

        // Step 1: Split the list into two halves
        Node[] split = splitList(head);
        Node left = split[0];
        Node right = split[1];

        // Step 2: Recursively sort each half
        left = mergeSort(left);
        right = mergeSort(right);

        // Step 3: Merge the sorted halves
        return merge(left, right);
    }

    // Function to split the list into two halves
    static Node[] splitList(Node head) {
        if (head == null || head.next == null) {
            return new Node[]{head, null};
        }

        Node slow = head;
        Node fast = head;
        Node prev = null;

        // Move fast by 2 and slow by 1
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        // Split the list into two halves
        prev.next = null;
        return new Node[]{head, slow};
    }

    // Function to merge two sorted lists
    static Node merge(Node left, Node right) {
        Node dummy = new Node(-1);
        Node current = dummy;

        while (left != null && right != null) {
            if (left.data <= right.data) {
                current.next = left;
                left = left.next;
            } else {
                current.next = right;
                right = right.next;
            }
            current = current.next;
        }

        // Append the remaining nodes
        if (left != null) {
            current.next = left;
        }
        if (right != null) {
            current.next = right;
        }

        return dummy.next;
    }
}

File: DSA/Arshdeep-Kaur/Day 56/Day56_Question2.cpp
class Solution {
  public:
    Node* findMid(Node* head) {
        Node* slow = head;
        Node* fast = head -> next;
    
        while(fast != NULL && fast -> next != NULL) {
            fast = fast -> next -> next;
            slow = slow -> next;
        }
    
        return slow;
    }
    
    Node* mergeLists(Node * left, Node* right) {
        //if left part is empty
        if (left == NULL) {
            return right;
        }
    
        //if right part is empty
        if (right == NULL) {
            return left;
        }
        
        Node* ans = new Node(-1);
        Node* temp = ans;
    
        //merge 2 sorted Linked List
        while (left != NULL && right != NULL) {
            if (left -> data < right -> data) {
                temp -> next = left;
                temp = left;
                left = left -> next;
            }
            else {
                temp -> next = right;
                temp = right;
                right = right -> next;
            }
        }
    
        while (left != NULL) {
            temp -> next = left;
            temp = left;
            left = left -> next;
        }
    
        while (right != NULL) {
            temp -> next = right;
            temp = right;
            right = right -> next;
        }
    
        temp = ans -> next;
        delete ans;
        return temp;
    }
    
    Node* mergeSort(Node* head) {
        //base case
        if (head == NULL || head -> next == NULL) {
            return head;
        }
    
        //break linked list into 2 halves, after finding the mid
        Node* mid = findMid(head);
    
        Node* left = head;
        Node* right = mid -> next;
        mid -> next = NULL;
    
        //recursive calls to sort both halves
        left = mergeSort(left);
        right = mergeSort(right);
    
        //merge both left and right halves
        Node* result = mergeLists(left, right);
    
        return result;
    }
};
